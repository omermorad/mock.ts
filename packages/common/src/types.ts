import faker from 'faker';
import FakerStatic = Faker.FakerStatic;

export type Faker = FakerStatic;
export const Faker = faker;

export interface ObjectLiteral {
  [key: string]: unknown;
}

export type Class<T = any> = new (...args: any[]) => T;

export type AbsoluteValue = string | number | boolean | ObjectLiteral | RegExp | Date;

export type FakerCallback = (faker: Faker) => any;

export type LazyType<T = any> = () => Class<T>;

export type Enum<E> = Record<keyof E, number | string> & { [key: number]: string };

export type LazyEnum<E = any> = { enum: () => Enum<E> };

export type ClassLiteral<TClass = any> = { [K in keyof TClass]: TClass[K] };

export type OptionalClassValues<TClass> = Partial<ClassLiteral<TClass>>;
