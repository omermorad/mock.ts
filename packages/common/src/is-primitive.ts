/**
 * Checks if a given constructor name is a primitive one
 *
 * @param constructorName
 * @returns {boolean}
 */
export function isPrimitive(constructorName: string): boolean {
  return ['String', 'Boolean', 'Number', 'Date'].includes(constructorName);
}
