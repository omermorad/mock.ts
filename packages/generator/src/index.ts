export * from '@mockinbird/parser';
export * from '@mockinbird/types';

export { Mock, MockOptions } from '@mockinbird/reflect';

export * from './lib';
export * from './types/mock-generator-options.interface';
