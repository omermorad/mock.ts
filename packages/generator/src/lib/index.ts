export * from './mock-factory';
export * from './builder';
