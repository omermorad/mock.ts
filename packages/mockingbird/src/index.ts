export { Mock, DecoratorValueArg } from '@mockingbird/reflect';

export * from '@mockingbird/parser';
export * from '@mockingbird/common';

export * from './lib';
