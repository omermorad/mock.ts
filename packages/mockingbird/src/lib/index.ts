export * from './factory/mock-factory';
export * from './builder';
