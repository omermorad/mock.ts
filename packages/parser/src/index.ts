import { Container } from 'typedi';
import { ClassParser } from './lib/parser/class-parser';
import { Faker } from '@mockinbird/common';
import RandExp from 'randexp';

export * from './types/types';
export { ClassParser };

Container.set('Faker', Faker);
Container.set('RandExp', RandExp);

export const classParser = Container.get(ClassParser);
