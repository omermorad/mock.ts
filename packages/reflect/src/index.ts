export * from './types';
export * from './decorators';
export * from './lib';
