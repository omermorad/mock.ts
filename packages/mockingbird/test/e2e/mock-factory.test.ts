import { MockFactory } from '../../src';
import { TestClassesE2E } from './test-classes';
import * as faker from 'faker';
import Bird = TestClassesE2E.Bird;

describe('MockFactory e2e Test', () => {
  let result;

  when("calling 'many' method with a given class", () => {
    beforeAll(() => {
      result = MockFactory<Bird>(Bird).many(3);
    });

    then('return an array with the exact same length', () => {
      expect(result).toBeInstanceOf(Array);
      expect(result).toHaveLength(3);
    });

    then('return 3 instances of the given class', () => {
      expect(result).toMatchSnapshot();
    });
  });

  scenario('using a plain object', () => {
    when('I create 3 dogs and covert them into a plain object', () => {
      beforeAll(() => (result = MockFactory(Bird).plain().many(3)));

      test('then return a nested plain object', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('using mutations', () => {
    let mock;

    when('I use mutation with the mock factory', () => {
      beforeAll(() => {
        mock = MockFactory(Bird).mutate({ name: 'Mutated Name' }).one();
      });

      test("return a mock where the name is 'Mutated Name'", () => {
        expect(mock).toMatchSnapshot();
        expect(mock.name).toBe('Mutated Name');
      });
    });

    when('I use mutations (using faker) with the mock factory', () => {
      let mock;

      beforeAll(() => {
        faker.name.firstName = () => 'FakedBirdName';
        mock = MockFactory(Bird)
          .mutate((faker) => ({ name: faker.name.firstName() }))
          .one();
      });

      test("return a mock where the name is 'FakedBirdName'", () => {
        expect(mock.name).toBe('FakedBirdName');
      });
    });
  });
});
