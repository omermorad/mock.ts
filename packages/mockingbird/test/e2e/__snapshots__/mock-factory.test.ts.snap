// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MockFactory e2e Test Scenario:  using a plain object when I create 3 dogs and covert them into a plain object then return a nested plain object 1`] = `
Array [
  Object {
    "name": "BirdyBird",
    "owner": Object {
      "car": Object {
        "model": "BMW",
      },
      "gender": "male",
    },
  },
  Object {
    "name": "BirdyBird",
    "owner": Object {
      "car": Object {
        "model": "BMW",
      },
      "gender": "male",
    },
  },
  Object {
    "name": "BirdyBird",
    "owner": Object {
      "car": Object {
        "model": "BMW",
      },
      "gender": "male",
    },
  },
]
`;

exports[`MockFactory e2e Test using mutations when I use mutation with the mock factory return a mock where the name is 'Mutated Name' 1`] = `
Bird {
  "name": "Mutated Name",
  "owner": Person {
    "car": Car {
      "model": "BMW",
    },
    "gender": "male",
  },
}
`;

exports[`MockFactory e2e Test when calling 'many' method with a given class then return 3 instances of the given class 1`] = `
Array [
  Bird {
    "name": "BirdyBird",
    "owner": Person {
      "car": Car {
        "model": "BMW",
      },
      "gender": "male",
    },
  },
  Bird {
    "name": "BirdyBird",
    "owner": Person {
      "car": Car {
        "model": "BMW",
      },
      "gender": "male",
    },
  },
  Bird {
    "name": "BirdyBird",
    "owner": Person {
      "car": Car {
        "model": "BMW",
      },
      "gender": "male",
    },
  },
]
`;
